smslibrary/src/main/java/com/eis/smslibrary/SMSMessage.java

```diff
@@ -38,16 +46,43 @@ public SMSMessage(@NonNull SMSPeer peer, @NonNull String messageText) throws Inv
     }
 
     /**
-     * Checks if the message content could be valid.
+     * Checks if the message content is valid.
      *
      * @param messageText to be checked.
-     * @return The state of the message after the validity tests.
+     * @return The state of the message after validity tests.
      */
-    public static ContentState checkMessageText(@NonNull String messageText) {
-        if (messageText.length() > SMSMessage.MAX_MSG_TEXT_LEN) {
-            return ContentState.MESSAGE_TEXT_TOO_LONG;
+    static ContentState checkMessageText(@NonNull String messageText) {
+        /* Since applications using this library will probably try to avoid wasting characters,
+         * they'll mostly stick to using the regular GSM character set, so we perform pattern
+         * matching on the whole message first, to avoid multiple checks on single characters
+         * when not necessary.
+         */
+        if (messageText.matches(GSM_CHARACTERS_STRING_REGEX)) {
+            // messageText contains only GSM characters
+            if (messageText.length() <= MAX_MSG_TEXT_LEN) {
+                return ContentState.MESSAGE_TEXT_VALID;
+            }
+        } else if (!messageText.matches(GSM_CHARACTERS_EXTENSION_STRING_REGEX)) {
+            // messageText contains Unicode characters
+            if (messageText.length() <= MAX_UTF16_MSG_TEXT_LEN) {
+                return ContentState.MESSAGE_TEXT_VALID;
+            }
+        } else {
+            // messageText contains only chars from both the GSM charset and its extension table
+            int charNum = 0;
+            for (int i = 0; i < messageText.length(); i++) {
+                String currentChar = messageText.substring(i, i + 1);
+                if (currentChar.matches(GSM_CHARACTERS_EXTENSION_REGEX)) {
+                    charNum += 2;
```

Comment:
Explain why + 2

Reply:
GSM characters from the extension table are counted as two characters, because when `messageText` is converted to GSM charset encoding, each of those special characters is translated to an escape character + the character from the extension table. See https://en.wikipedia.org/wiki/GSM_03.38#GSM_7-bit_default_alphabet_and_extension_table_of_3GPP_TS_23.038_/_GSM_03.38

Reply:
OK